Create issue-triage-process.md (#2811)
kubeapps/kubeapps: docs/developer/issue-triage-process.md

# Kubeapps triage process

## Kubeapps backlog

> Kubeapps keeps a backlog of issues on GitHub submitted both by maintainers and contributors: this backlog comprises bugs, feature requests, and technical debt. If it’s a bug, an idea for a new feature, or something in between, it’s filed as an issue in the [kubeapps issue page](https://github.com/kubeapps/kubeapps/issues) on GitHub.

Reviewer: Not sure about `’` vs `'`.

Reviewer 2: https://guides.github.com/features/mastering-markdown/ See the examples for quotes there. So far we've always just used the normal `'` and let GH/Markdown smart quotes handle it (whether or not it does that correctly).

There are some special considerations about how the Kubeapps maintainer team manages its backlog on GitHub:
- The issue repository is completely open. The maintainer team, along with the entire Kubeapps community, files all feature enhancements, bugs, and potential future work into the open repository.
- Issues are closed if solved, or outdated (meaning the issue does not apply anymore according to the evolution of Kubeapps or it was waiting for more information which was never received).

## Kubeapps' triage process

The [kubeapps/kubeapps](https://github.com/kubeapps/kubeapps) issues are triaged by any member of the maintainer team. The triage process is defined with the objective of helping the Kubeapps team during the planning of each iteration so that the selected issues provide the highest added value to the product itself according to the Kubeapps maintainer team capacities.

> The triage process should be performed manually for any issue in the Inbox column and will consist on:

Reviewer 2: ```suggestion
The triage process should be performed manually for any issue in the Inbox column and will consist of:
```

- Reading the description of the issue. 
  * If more information is requested, as a result of the triage process the issue must be labeled as `<awaiting-more-evidence>`, a comment requesting for information added and the issue must remain in the Inbox column.
- Adding labels for each category (required): component, kind, size and priority.

> - Check if it is a `good-first-issue` to start contributing to Kubeapps and label the issue in consequence.

Reviewer 2: ```suggestion
- Check if it is a `good-first-issue` to start contributing to Kubeapps and label the issue as such.
```

An extra point which is often part of the process:

- Investigate whether the issue is reproducible in a environment and update with specific instructions. Sometimes this investigation may lead to an obvious, simple fix which can be proposed right away but depending on the priority it may instead be tagged as a `good-first-issue` as above.

> - Moving the issue to the appropriate column according within the triage process:

Reviewer: s/according within/according to  ?

  - **Committed** → Issues labeled as `priority/unbreak-now`
  - **Next** iteration discussion → Issues labeled as `priority/high`

>   - **Inbox** → Issues labeled `priority/medium` and `priority/low` will remain in the inbox before moving to the **Backlog** to double check with the whole maintainer team during the iteration planning meeting.

Reviewer: s/double check/double-check

  - **Inbox** → Issues labeled as `awaiting-more-evidence`
- Repeat until all issues in the **Inbox** column are triaged.

Besides this manual triage process, there are two automatic process lead by the bots configured in GitHub:
+ Automatically move new and reopened issues to the **Inbox** column to start the triage process (due to eventual project-bot outages, once a month we should manually search for issues with "no:project" assigned).

> + Automatically **dependabot** and **Snyk** are labelling PRs depending on the target language.

Reviewer: Question: do we prefer en_gb rather than en_us ? I guess so, but perhaps we should check the corporate guidelines to shed some light on it.

+ Automatically **stalebot** is checking inactive issues to label them as `stale`. An issue becomes stale after 60 days of inactivity and the bot will close it after 7 days of inactivity for stale issues. To be considered:

>   * Issues labeled as `priority/unbreak-now` → `priority/high` → `priority/medium` will never be labeled as `stale`.

Reviewer 2: I wasn't sure if you'd made `kind/bug` also exempt from going stale?

Author: Right. I have to add it.

`M`|A story that can be done by a person in 4-7 days, possibly with some uncertainty
`L`|A story that requires investigation and possibly will take a person a full 2-week iteration
`XL`|A story too big or with too many unknowns. Needs investigation and split into several ones

### Priority

priority/| | | |
--------|-|-|-|
`unbreak-now`|`high`|`medium`|`low`

**Contribution labels**:

`awaiting-more-evidence` - information requested to the reporter.

`help-wanted` - the maintainer team wants help on an issue or pull request.

`good-first-issue` - good first issues to start contributing to Kubeapps.

Automatic labels| | | | | |
----------------|-|-|-|-|-|
`dependencies`|`go`|`javascript`|`rust`|`security`|`stale`

## Kubeapps milestones
Next step in the triage process to help the community understand the project roadmap is to define milestones and associate issues to them.
According to the Kubeapps team practices, milestones will be defined based on quarters and years according to the following pattern: **YYYY-QN** (i.e. 2021-Q1 - 2021-Q2 - etc), setting the due date for each milestone.

## Kubeapps iteration planning guidelines

> 1. Review issues in column **Inbox** (untriaged issues, awaiting-more-evidence, triaged issues labeled as low and medium priority) and move them to the column according to the triage process. Issues labeled as `awaiting-more-evidence` must be checked if updated to be triaged.

Reviewer 2: Perhaps it's not relevant here as it's not part of triaging, but we'd normally go through the current in-progress work to get an update so that we also have a better idea of how much we can commit to for the next iteration.
